From f73813a52536bb7501c71e74613469fb792dd655 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:38:57 -0400
Subject: [net] etherdevice: Use ether_addr_copy to copy an Ethernet address

Message-id: <1409658042-22689-177-git-send-email-sassmann@redhat.com>
Patchwork-id: 90371
O-Subject: [RHEL7.1 PATCH 176/281] etherdevice: Use ether_addr_copy to copy an Ethernet address
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

From: Joe Perches <joe@perches.com>

Some systems can use the normally known u16 alignment of
Ethernet addresses to save some code/text bytes and cycles.

This does not change currently emitted code on x86 by gcc 4.8.

Signed-off-by: Joe Perches <joe@perches.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 286ab723d4b83d37deb4017008ef1444a95cfb0d)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/etherdevice.h b/include/linux/etherdevice.h
index c623861..9ff0c77 100644
--- a/include/linux/etherdevice.h
+++ b/include/linux/etherdevice.h
@@ -199,6 +199,28 @@ static inline void eth_hw_addr_random(struct net_device *dev)
 }
 
 /**
+ * ether_addr_copy - Copy an Ethernet address
+ * @dst: Pointer to a six-byte array Ethernet address destination
+ * @src: Pointer to a six-byte array Ethernet address source
+ *
+ * Please note: dst & src must both be aligned to u16.
+ */
+static inline void ether_addr_copy(u8 *dst, const u8 *src)
+{
+#if defined(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)
+ *(u32 *)dst = *(const u32 *)src;
+ *(u16 *)(dst + 4) = *(const u16 *)(src + 4);
+#else
+ u16 *a = (u16 *)dst;
+ const u16 *b = (const u16 *)src;
+
+ a[0] = b[0];
+ a[1] = b[1];
+ a[2] = b[2];
+#endif
+}
+
+/**
  * compare_ether_addr - Compare two Ethernet addresses
  * @addr1: Pointer to a six-byte array containing the Ethernet address
  * @addr2: Pointer other six-byte array containing the Ethernet address
-- 
1.7.1